pkg/cluster/bootstrap.go

```diff
@@ -85,47 +90,219 @@ func (c *Cluster) shouldBootstrapLoad(ctx context.Context) (bool, error) {
+// to the older.
+func (c *Cluster) ReconcileBootstrapData(ctx context.Context, crb *config.ControlRuntimeBootstrap) error {
+	logrus.Info("Reconciling bootstrap data between database and disk")
+
+	storageClient, err := client.New(c.etcdConfig)
+	if err != nil {
+		return err
+	}
+
+	token := c.config.Token
+	if token == "" {
+		tokenFromFile, err := readTokenFromFile(c.runtime.ServerToken, c.runtime.ServerCA, c.config.DataDir)
+		if err != nil {
+			return err
+		}
+		if tokenFromFile == "" {
+			// at this point this is a fresh start in a non-managed environment
+			c.saveBootstrap = true
+			return nil
+		}
+		token = tokenFromFile
+	}
+	normalizedToken, err := normalizeToken(token)
+	if err != nil {
+		return err
+	}
+
+	var value *client.Value
+
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for range ticker.C {
+		value, err = c.getBootstrapKeyFromStorage(ctx, storageClient, normalizedToken, token)
+		if err != nil {
+			if strings.Contains(err.Error(), "not supported for learner") { // create error value
+				continue
+			}
+			return err
+		}
+		if value == nil {
+			return nil
+		}
+
+		break
+	}
+
+	data, err := decrypt(normalizedToken, value.Data)
+	if err != nil {
 		return err
 	}
 
-	// return if file already exists
-	if _, err := os.Stat(stamp); err == nil {
-		return nil
+	if err := c.certDirsExist(); err != nil {
+		logrus.Warn(err.Error())
+		return bootstrap.WriteToDiskFromStorage(bytes.NewBuffer(data), crb)
 	}
 
-	// otherwise try to create it
-	f, err := os.Create(stamp)
+	paths, err := bootstrap.ObjToMap(crb)
 	if err != nil {
 		return err
 	}
 
-	return f.Close()
+	files := make(bootstrap.DataFormat)
+	if err := json.NewDecoder(bytes.NewBuffer(data)).Decode(&files); err != nil {
+		// This will fail if data is being pulled from old an cluster since
+		// older clusters used a map[string][]byte for the data structure.
+		// Therefore, we need to perform a migration to the newer bootstrap
+		// format; bootstrap.BootstrapFile.
+		if err := c.migrateBootstrapData(ctx, storageClient, data, files, normalizedToken, value.Modified); err != nil {
+			return err
+		}
+	}
+
+	for pathKey, fileData := range files {
+		path, ok := paths[pathKey]
+		if !ok {
+			continue
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		fd, err := ioutil.ReadAll(f)
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(fileData.Content, fd) {
+			logrus.Warnf("%s is out of sync with datastore", path)
+
+			info, err := os.Stat(path)
+			if err != nil {
+				return err
+			}
+
+			switch {
+			case info.ModTime().Unix() > files[pathKey].Timestamp.Unix():
+				logrus.Warn(info.Name() + " newer than within database")
+				return c.save(ctx, true)
+			case info.ModTime().Unix() < files[pathKey].Timestamp.Unix():
+				logrus.Warn("database newer than " + info.Name())
+				return bootstrap.WriteToDiskFromStorage(bytes.NewBuffer(data), crb)
```

Comment:
Hmm, so if a single local file is newer we write to the datastore, and if a single file from the bootstrap data is newer we write to disk? I feel like we need to individually reconcile files depending on which is newer, rather than syncing all files based on the first finding of one that doesn't have a matching timestamp.

Reply:
I am not sure if that's needed, since we store all bootstrap data as one blob and not separately for each file then I think its fine to write to storage all certs

Reply:
right, but I'm not sure that's the right thing to do. What if one cert newer in the bootstrap data, another is newer on the local disk, and the rest are the same. How should that be handled?

Reply:
This seems like something that we can handle with documentation. ie "Perform certificate operations on the sane host to guarantee successful replication", or something to that effect.

Reply:
How about at least **checking for** and warning if file modifications would conflict, instead of taking action as soon as we find a single file? Scan the whole file list and set flags indicating if the local files are newer/older. If we find both newer and older then warn or error. That also gives us the opportunity to write log entries for any files that will be touched by the sync.

Reply:
updated